### PR TITLE
Payments refactor

### PR DESCRIPTION
**Pull request by diogob**
* Creates Payment model
 * Moves contribution state machine to payment
 * Uses ContributionDetail view in admin and moves some scopes from Contribution
 * Refactor catarse_pagarme 
 * Refactor contribution creation

### CODE REVIEW

File: app/controllers/admin/contributions_controller.rb [hunk 1]
```diff
@@ -30,7 +29,11 @@ def set_title
     @title = t("admin.contributions.index.title")
   end
 
+  def resource
+    Payment.find params[:id]
```
Comment (adrianob) [inline app/controllers/admin/contributions_controller.rb]: Why is the resource a Payment and not a Contribution?

Comment (diogob) [inline app/controllers/admin/contributions_controller.rb]: Yeah, this is a bit weird. But we are not using the contribution here anymore. The collection method now uses the ContributionDetail which is basically ```contributions JOIN payments```. I used the Payment as a resource so I would not have to proxy every Payment method through ContributionDetail.

File: app/decorators/contribution_decorator.rb [hunk 2]
```diff
@@ -25,12 +25,29 @@ def display_value
...
   def display_slip_url
-    return source.slip_url if source.slip_url.present?
-    "https://www.moip.com.br/Boleto.do?id=#{source.payment_id.gsub('.', '').to_i}"
+    gateway_data = last_payment.try(:gateway_data)
+    return gateway_data["boleto_url"] if gateway_data.present?
+  end
+
+  def display_status
+    if last_payment.pending?
+      "Aguardando confirmação do pagamento"
```
Comment (vicnicius) [inline app/decorators/contribution_decorator.rb]: Shouldn't we make those status translatable?

Comment (diogob) [inline app/decorators/contribution_decorator.rb]: Good catch, I refactored the spec, and forgot to finish the method :P

File: app/models/contribution.rb [hunk 3]
```diff
@@ -4,40 +4,28 @@ class Contribution < ActiveRecord::Base
...
   scope :anonymous, -> { where(anonymous: true) }
-  scope :credits, -> { where("credits OR lower(payment_method) = 'credits'") }
   scope :not_anonymous, -> { where(anonymous: false) }
-  scope :confirmed_today, -> { with_state('confirmed').where("contributions.confirmed_at::date = to_date(?, 'yyyy-mm-dd')", Time.now.strftime('%Y-%m-%d')) }
-  scope :confirmed_last_day, -> { with_state('confirmed').where("confirmed_at > current_timestamp - interval '1 day'") }
+  scope :confirmed_last_day, -> { where("EXISTS(SELECT true FROM payments p WHERE p.contribution_id = contributions.id AND p.state = 'paid' AND (current_timestamp - p.paid_at) < '1 day'::interval)") }
+  scope :pending, ->{
+    where("EXISTS (SELECT true FROM payments p WHERE p.contribution_id = contributions.id AND p.state = 'pending') AND NOT contributions.was_confirmed")
+  }
   scope :avaiable_to_automatic_refund, -> {
```
Comment (adrianob) [inline app/models/contribution.rb]: Typo here.

Comment (adrianob) [inline app/models/contribution.rb]: Typo on scope name.

File: app/models/contribution.rb [hunk 4]
```diff
@@ -80,6 +67,10 @@ def can_refund?
     confirmed? && project.failed?
   end
 
+  def confirmed?
+    @confirmed ||= Contribution.where(id: self.id).pluck('contributions.is_confirmed').first
```
Comment (adrianob) [inline app/models/contribution.rb]: Why not simply use find instead of where?

**Comment by diogob** [comment]
@devton any remarks here?

**Pull request merged**